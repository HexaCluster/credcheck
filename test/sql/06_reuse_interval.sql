DROP USER IF EXISTS credtest;
DROP EXTENSION credcheck CASCADE;
CREATE EXTENSION credcheck;
-- no password in the history, the extension has not been loaded
CREATE USER credtest WITH PASSWORD 'AJ8YuRe=6O0';
LOAD 'credcheck';
SET credcheck.password_reuse_history = 1;
SET credcheck.password_reuse_interval = 365;
SELECT rolename, password_text FROM credcheck.pg_auth_history ;
-- Add a new password in the history and set its age to 100 days
ALTER USER credtest PASSWORD 'J8YuRe=6O';
UPDATE credcheck.pg_auth_history SET password_date = now() - '100 days'::interval;
SELECT rolename, password_text FROM credcheck.pg_auth_history ;
-- fail, the password is in the history for less than 1 year
ALTER USER credtest PASSWORD 'J8YuRe=6O';
SELECT rolename, password_text FROM credcheck.pg_auth_history ;
-- success, but the old password must be kept in the history (interval not reached)
ALTER USER credtest PASSWORD 'AJ8YuRe=6O0';
SELECT rolename, password_text FROM credcheck.pg_auth_history ;
-- fail, the password is still present in the history
ALTER USER credtest PASSWORD 'J8YuRe=6O';
-- Change the age of the password to exceed the 1 year interval
UPDATE credcheck.pg_auth_history SET password_date = now() - '380 days'::interval;
-- success, the old password present in the history has expired
ALTER USER credtest PASSWORD 'J8YuRe=6O';
SELECT rolename, password_text FROM credcheck.pg_auth_history ;
-- Rename user, all entries in the history table must follow the change
ALTER USER credtest RENAME TO credtest2;
SELECT rolename, password_text FROM credcheck.pg_auth_history ;
-- Dropping the user must empty the record in history table
DROP USER credtest2;
SELECT * FROM credcheck.pg_auth_history;
