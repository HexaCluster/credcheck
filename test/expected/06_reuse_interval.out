DROP USER IF EXISTS credtest;
NOTICE:  role "credtest" does not exist, skipping
DROP EXTENSION credcheck CASCADE;
CREATE EXTENSION credcheck;
-- no password in the history, the extension has not been loaded
CREATE USER credtest WITH PASSWORD 'AJ8YuRe=6O0';
LOAD 'credcheck';
SET credcheck.password_reuse_history = 1;
SET credcheck.password_reuse_interval = 365;
SELECT rolename, password_text FROM credcheck.pg_auth_history ;
 rolename | password_text 
----------+---------------
(0 rows)

-- Add a new password in the history and set its age to 100 days
ALTER USER credtest PASSWORD 'J8YuRe=6O';
UPDATE credcheck.pg_auth_history SET password_date = now() - '100 days'::interval;
SELECT rolename, password_text FROM credcheck.pg_auth_history ;
 rolename |                          password_text                           
----------+------------------------------------------------------------------
 credtest | e61e58c22aa6bf31a92b385932f7d0e4dbaba24fa3fdb2982510d6c72a961335
(1 row)

-- fail, the password is in the history for less than 1 year
ALTER USER credtest PASSWORD 'J8YuRe=6O';
ERROR:  Cannot use this credential following the password reuse policy
SELECT rolename, password_text FROM credcheck.pg_auth_history ;
 rolename |                          password_text                           
----------+------------------------------------------------------------------
 credtest | e61e58c22aa6bf31a92b385932f7d0e4dbaba24fa3fdb2982510d6c72a961335
(1 row)

-- success, but the old password must be kept in the history (interval not reached)
ALTER USER credtest PASSWORD 'AJ8YuRe=6O0';
SELECT rolename, password_text FROM credcheck.pg_auth_history ;
 rolename |                          password_text                           
----------+------------------------------------------------------------------
 credtest | e61e58c22aa6bf31a92b385932f7d0e4dbaba24fa3fdb2982510d6c72a961335
 credtest | 79320cea69ba581d5e17255c02ae08060f412f79a7c14d0e24ffca51fc03ec74
(2 rows)

-- fail, the password is still present in the history
ALTER USER credtest PASSWORD 'J8YuRe=6O';
ERROR:  Cannot use this credential following the password reuse policy
-- Change the age of the password to exceed the 1 year interval
UPDATE credcheck.pg_auth_history SET password_date = now() - '380 days'::interval;
-- success, the old password present in the history has expired
ALTER USER credtest PASSWORD 'J8YuRe=6O';
SELECT rolename, password_text FROM credcheck.pg_auth_history ;
 rolename |                          password_text                           
----------+------------------------------------------------------------------
 credtest | 79320cea69ba581d5e17255c02ae08060f412f79a7c14d0e24ffca51fc03ec74
 credtest | e61e58c22aa6bf31a92b385932f7d0e4dbaba24fa3fdb2982510d6c72a961335
(2 rows)

-- Rename user, all entries in the history table must follow the change
ALTER USER credtest RENAME TO credtest2;
SELECT rolename, password_text FROM credcheck.pg_auth_history ;
 rolename  |                   password_text                   
-----------+---------------------------------------------------
 credtest2 | 9320cea69ba581d5e17255c02a
 credtest2 | 61e58c22aa6bf31a92b385932f7d0e4dbaba24fa3fdb29825
(2 rows)

-- Dropping the user must empty the record in history table
DROP USER credtest2;
SELECT * FROM credcheck.pg_auth_history;
 rolename | password_date | password_text 
----------+---------------+---------------
(0 rows)

